### DISCUSSION TITLE
Standardization and best practices

### DISCUSSION THREAD

**Discussion by samikanza** [MADICES/MADICES-2025#1] [Ideas]
**MADICES 2024 Recap**
We delved into the realm of semantic annotation, emphasizing the adoption of RDF serialization formats for data representation, using JSON-LD as a working example, to enhance the interpretability and reusability of research datasets. Participants explored methodologies for incorporating metadata, handling missing values, and standardizing units for measurements. [Tools and guidelines](https://github.com/MADICES/semantic-annotation) were developed to streamline the annotation process and ensure compatibility across different knowledge domains.

**MADICES 2025 Focus Area**
We want to focus on further refinement of protocols and guidelines for semantic annotation and interoperable data exchange. Recent discussions with members at the Future Labs Live 2024 event in Basel illuminated the interest in semantic interoperability of research data and the need for standards and tooling to facilitate ease-of-use and reduce the burden/barrier to semantically annotate datasets. 

Please include links, comments, and discuss plans for this focus area below.

**Comment by hampusnasstrom** [comment]
There is a lot of focus on the semantic web with annotated data sets serialized as RDF, JSON-LD, or XML. However, I don't think we can expect every scientist to search through ontologies and connect them. This is rather a task for data stewards but the unfortunate reality is that most groups don't have access to one, and won't for the foreseeable future. I therefore believe that we need a way to **F**ind, **I**nteroperate, **A**ccess, and **R**euse not only the definitions but the collection of the these in schemas/forms.

**Reply by edan-bainglass** [reply]
> I don't think we can expect every scientist to search through ontologies and connect them

This burden is likely a non-starter for most researchers. Too busy doing research and writing papers :) Key is to find a way to integrate these tasks seamlessly into research workflows. At AiiDA, we are looking into integrating semantics directly into our data and logic pipelines, such that any research carried out via the AiiDA engine is automatically semantic!

One challenge, as you've pointed out, is which ontologies to use? How to handle "redundant" ontologies, i.e., reinvented ontologies (due perhaps to lack of awareness of other works) or mildly different ontologies? How do we map these without burdening users, such that they are easily connected to the "global research knowledge"?

We've been discussing internally how LLMs might assist in such tasks. AI data stewards? Surely this would be discussed at Future Labs Live (I believe it was already in 2024). Best to keep an eye on this and try to drag a few experts to MADICES :)

**Reply by simontaurus** [reply]
We made progress in the [oold-python](https://github.com/OO-LD/oold-python) implementation in order to keep all semantics in pydantic dataclasses you may just import from a common repo:

```python
# Example: use
import Person from opensemantic.base

p1 = Person(name="Alice")
p2 = Person(name="Bob", knows=[p1])

# Export to JSON-LD / RDF
json_ld = p2.to_jsonld())

# Import JSON-LD / RDF
p3 = Person.from_jsonld(json_ld)

# access a linked entity directly via property chain
print(p3.knows[0].name) # Alice

# inline query
print(p3.knows[Person.name == "Alice"])

...
```

**Comment by PeterKraus** [comment]
One of the goals I have for MADICES-2025 is to understand how do I need to annotate my NetCDF datasets so that the annotations are useful downstream. By this I mean the mechanics of it, not figuring out the ontology labels etc.

I am happy to provide some example files, and prepare a spreadsheet matching data headers to ontology entries. These annotations could be implemented into either [`yadg`](https://dgbowl.github.io/yadg/main/index.html) or [`tomato`](https://dgbowl.github.io/tomato/main/index.html).

**Reply by PeterKraus** [reply]
A quick summary of the discussions I've had with folks at MADICES over these weeks:

## No clear standard
There are several well-understood ways to semantically annotate **metadata**. These might not be compatible with each other, for technical reasons (the various container formats used to talk to each other; various functionalities such as the while-loop issue in workflows) and "social" reasons (incompatible base ontologies), but at least there is a **significant amount** of previous work.

For annotation of research **data** as stored within the files we'd like to pass, nobody that I talked to at MADICES could point me to a filetype-agnostic standard. Just to be absolutely clear: this is about how column and row labels in tabular data, or coordinate labels in n-dimensional data, should be annotated, where should they be attached to, and in what format. Ontology type/term look-up is out of scope as it's well covered by other efforts at MADICES and otherwise.

## Prior work
- [Frictionless data](https://frictionlessdata.io/) with its [Data Package](https://datapackage.org/standard/data-package/) standard seems to be closest to a "filetype agnostic" standard and might be the best way forward - if not directly, at least as an inspiration. Abigail Miller remarked support for multi-dimensional data and hierarchical groups (xarray, hdf5) is most likely currently missing.
- @NukP mentioned/showed a way to annotate the content of parquet files as part of their battery ontology work, that has been developed together with @jsimonclark; the recommendation was to contact Simon for more info.
- The Herbie platform which I talked to @martin-held-hereon about is fully semantically annotated. Unfortunately, it's used to store "laboratory notebook" entries in the original sense as opposed to digitized and annotated scientific data files.
- There is the [HDF5-based RDM toolbox](https://h5rdmtoolbox.readthedocs.io/en/latest/) and of course FAIRmat's work on [NeXus](https://fairmat-nfdi.github.io/pynxtools/) proliferation and adoption
- Discussions with several people attending MADICES-2025 revealed doubts on whether such annotation is at all worth it.

## Wishlist
While no agreement could be reached, I am more optimistic and see a way forward, at least in my work, to enable semantic annotations. The following criteria / recommendations should be adhered to for maximum compatibility with existing prior art and developments in other areas:

- an honest effort should be made to make any such annotation **compatible with base RO-crate** standard or profile
- the manifest of the root RO-crate (`ro-crate-metadata.json`) should **not** be "polluted" with data annotations; ideally each data file within the RO-crate dataset would be annotated using a separate metadata file, including a self-describing and self-consistent schema, indicated using `conformsTo` or similar
- only BFO-compatible ontologies should be used for compatibility with existing state-of-the-art
- definition of new ontologies or terms should be avoided
- has to work with the concept of hdf5 groups (or similar)
- in order to allow semantic annotation of "legacy" (pre-existing) data, no restriction on the file format used to store the data should be made, provided the format itself is an open standard (e.g. hdf5, sqlite dump, pandas pickle, json, csv, ...) i.e. not a vendor-specific or legacy binary format, even if an extractor exists; note that raw data in vendor/proprietary formats should be stored, but semantic annotation of those files is of doubtful value given that the data within cannot be accessed without proprietary tools
- the previous point implies data conversion to a higher-dimensional format (pandas -> xarray) should not be necessary
- known, existing efforts in this area should be considered

I had also made a personal request that this topic be revisited at MADICES 4.

## Next steps
If time constraints allow, I would be happy to work on an implementation of the above principles for my own research data, first by exploring the Frictionless Data approach further. If you're keen to contribute or feel like I missed anything, please let me know here or in private.

**Reply by ml-evs** [reply]
See https://github.com/MADICES/MADICES-2025/discussions/32 for @scman1's take on this (and perhaps a place you @PeterKraus could collaborate on c.f. making the results of the ongoing CDIF-XAS project suit your needs)

**Reply by SteffenBrinckmann** [reply]
Thanks for the recap. 

1. What is the reason for considering additional metadata as pollution? I would call it one consistent dataset graph in which branches can be cropped depending on need. Since the graphs are flattened anyhow in ro-create, the difference is just one long file vs multiple shorter files. The single file with its shasums of datafiles can be easily signed allowing to verify content. The only downside comes if the file becomes too big. What are your reasons for a multitude file approach?

2. For my version of good scientific conduct (provenance), possibly increased metadata, possibly increased accuracy and hindering scientific fraud, I would always strongly state, save the original binary data in all cases (even if no converter exists). It does not significantly hurt the file size and has too many potential upsides.
But you and Tim Berners Lee LOD disagree. Does it have technical reasons or more philosophical?

**Reply by PeterKraus** [reply]
Good comments.

1. I believe it's about separation of roles. There was a clear preference from Juan and Abigail that `ro-crate-metadata.json` should focus on describing the layout of the RO-crate, and the datafiles should be annotated outside the main manifest (if at all). Additionally, one might want to use a pre-existing standard (like Frictionless Data or CSVW) to annotate the data files themselves, or use standard-compliant NeXus files which don't need annotation, while RO-crate is JSON-LD for the whole dataset.
2. This is a misunderstanding! Raw data should always be stored! However, there is no point doing semantic annotations of data in files which cannot be opened without additional proprietary software. It should be converted / exported into an open file format first.

**Reply by SteffenBrinckmann** [reply]
Is the separation really as clear as possible? A datafile has all its annotations, using Nexus, CSVW or FLD, and the ro-crate should not mirror it because data-duplicity always carries the risk of differences. The crate manifest has the layout. But where goes the grey stuff: comments about the measurement, link to sample/procedure/device, etc.? If the next room has shown a broken temperature control, that context might influence the present measurements and this context should be embedded in the manifest,json as it shows the context. 
IMHO, the boundary should be: did the instrument vendor / creator use this (meta)data -> the metadata should be in the Nexus, CSVW, etc. file. If the (meta)data comes from the context of the measurement, it should be in the metadata.json.

Does anybody know one or multiple good examples that can be discussed such that one come to a "gold standard" (as set of minimal properties everybody agrees to) with lots of optional properties (that some people object)? At least, having such an example makes things explicit and prevents some misunderstandings like mine about the raw-files.

**Reply by PeterKraus** [reply]
@SteffenBrinckmann what you're describing is a problem, but it is out of scope here. This thread is strictly about describing the data (and their headers, attributes, types, units, etc.) as stored in "data files", using semantic annotations. 

Putting the annotations (whether they're complete or not) "somewhere" into a HDF5 file doesn't solve the issue, as I still need to indicate where that schema is stored and how can other codes access it. Is it on the root group? Is it on each group? Is it on each `data_var` and `coord`? Do I use JSON-LD syntax (`@type`, `@id` and friends) within HDF5? If you can point me to a prior art that does this for multi-dimensional data (except NeXus), I'd be happy to have a look.

I will point out that CSVW (which basically solves this for tabular data) is basically a JSON-LD annotation of a CSV file, so not a single file; I imagine one would indicate that these two should be consumed "together" using e.g. `conformsTo` parameter on the `File` (or whatever) node within the RO-crate.

**Reply by SteffenBrinckmann** [reply]
Sorry, that boundary of the scope was unclear to me as I took the first bullet point as indicating ro-crate is in scope. 

I also have been trying to find an alternative for hdf5/Nexus for a while because of the issues (instability for edit of files and ambiguity of unit location) and even had call with core team. Evan inside Nexus there is ambiguity, which results in lots of if statements for the consumer. 

If you find something, let me know, I will take a look if it is not too exotic.

**Reply by SteffenBrinckmann** [reply]
https://www.nature.com/articles/s41592-021-01326-w

**Reply by simontaurus** [reply]
Regarding `ro-crate-metadata.json` and additional files:
We should see RO-Crate as the RDF graph it is and from this perspective it is not relevant if we spread it over multiple files or not **if** the other files are (i) itself RDF (serialized, e.g., as JSON-LD), (ii) contain a single root entity (not a `@graph`) and are (iii) proper linked (e.g., by `conformsTo`).
IMHO we should put as much possible inside `ro-crate-metadata.json` since links to other files may introduce potential errors and cause latency. Arguments against are scaling issues if the files become huge. Ownership of different files by different software parts is not really a point since you could simply merge the at the final state of packaging.
Finally the RO-Crate RDF graph should be as large as possible, containing also schemas of files, and, at least on-demand (contolled by following a file link or applying a dedicated JSON-LD @context) also actual data points if possible.
A significant advantace of CSVW over frictionless it that CSVW is already well define (and W3C standardized) RDF so you can seemlessly integrate it into the RO-Crate RDF graph.
Now imagin you throw millions of RO-Crate on a triple store and then are able to run a SPARQL query like "Give me all datasets containing a tabular dataset containing at least the columns A, B" (e.g., stress over strain)

**Reply by simontaurus** [reply]
Regarding higher dimensional data, hdf5 & co
The main reasons for these data formats is performance, compression and partial access on the price of accessibility and semantics.
It is much easier to model higher dimensional data conceptional as arrays of (potential) nested JSON objects even if you store them differently.
> Consider a property temperature, which is the temperature measured at different sea depths at different times. Hence, it has as domain a coordinate system with two axis; depth and time.

You could simple store the numbers in a matrix
```
[0, 0, 305]
[0, 100, 299]
[0, 200, 296]
[3600, 0, 303]
[3600, 100, 297]
[3600, 200, 294]
```
leaving the need to explain what they mean or you put them into a object-oriented structure

```yaml
type: ex:TimeResolvedSeriesOfDepthResolvedTemperature
data:
  - time: {value: 0}
    data: 
      - {type: DepthResolvedTemperature, depth: {value: 0, unit: ...}, temperature: {value: 305}}
      - {depth: {value: 100}, temperature: {value: 299}}
      - {depth: {value: 200}, temperature: {value: 296}}
  - time: {value: 3600}
    data: 
      - {depth: {value: 0}, temperature: {value: 303}}
      - {depth: {value: 100}, temperature: {value: 297}}
      - {depth: {value: 200}, temperature: {value: 294}}
```
which can easily be described by JSON-LD context and JSON-SCHEMA / Pydandic Dataclass (also storing default units and types to be applied).
see also: https://github.com/OO-LD/schema/issues/17#issuecomment-2856992072

Again, this would not only allow you interpret the data seamlessly but also allows you to query (in future advanced data repos) for datasets that contain (as some part) well defined types of data slices / characteristics.

**Comment by ml-evs** [comment]
My aim is similar to Peter's, how can I best annotate public or published entries (e.g. samples, devices, their relationships and attached measurements) in [datalab instances](https://github.com/datalab-org) so that it can be used downstream, even if semantic annotations are missing or incomplete.

**Comment by samikanza** [comment]
We are also keen to discuss interoperability between ELNs! There are clearly a few different efforts here, on the open source side there is the ELN Consortia, with the ELNFileFormat, and I'm seeing a lot of discussions about this at the Industry based conferences that I go to about the Allotrope Foundation and various companies working with the large-scale commercial ELNs to provide an archive format to pull data between them. 

Also interested in discussing common templates across ELNs

**Reply by cbarillari** [reply]
I am also very much interested in this topic. In the openBIS team we have done some work in this direction and our developers could contribute to discussions on this topic.

**Reply by edan-bainglass** [reply]
We need to understand precisely where the .eln file format was lacking leading us at PREMISE to pursue a custom extension of the RO Crate spec. It is important that we connect our PREMISE team with that of the ELN Consortium (particularly @SteffenBrinckmann), so that we sync up efforts and avoid confusion. @cbarillari great if you can ping your team with this topic, so they may comment.

**Comment by bdeadman** [comment]
We ([https://github.com/open-reaction-database](https://github.com/open-reaction-database), ORD) are interested in this topic. The ORD is an open-access repository and schema for chemical reaction data, particular experimental data of small molecule reactions. We use the [Protocol Buffers](https://protobuf.dev/) (similar to XML) for storing and exchanging reaction datasets, but also provide an Object Relational Mapper to unpack the data into a postgreSQL database.

Our interest is in working with other reaction data formats to ensure interoperability, but also thinking how we could fit into upstream and downstream data standards, and applications.

**Reply by edan-bainglass** [reply]
Pinging @fabioacl from the PREMISE team. One of our work packages addressed authoring semantically annotated schemas (and standards for doing so) for microscopy data (see [here](https://github.com/ord-premise/metadata-spectroscopy)), particularly studying molecules on surfaces. The schemas are being used to demonstrate interoperability between ELNs (openBIS in this case), where experimental data/metadata is being stored in the developed standards, and simulation engines (AiiDA in this case), where STM simulations are carried out to elucidate experimental results.

A key deliverable for us is the demonstration of data/metadata exchange across these platforms, for example:
- Lab results to openBIS
- openBIS objects to AiiDA for simulation
- AiiDA simulation results back to openBIS
- openBIS/AiiDA data/metadata to public archives

We also do something similar for battery research (another PREMISE work package), where battery testing experiments are orchestrated by a workflow engine, the data of which is exchanged with the openBIS ELN, again in PREMISE-developed standards/protocols. Pinging @NukP.

In both cases, we are exploring two paths:
1. Direct one-to-one interoperability using dedicated APIs from both platforms
2. A generalized platform-agnostic means of exchanging resources by means of RO Crates

Happy to share more here, in upcoming events, and of course at the workshop for those interested :)

**Comment by juan-fuentes-sis** [comment]
We really need a dedicated working group focused on advancing standardization efforts to improve the interoperability of systems, particularly with respect to data formats and associated tools.

As many of you have already mentioned, either directly or indirectly, we are facing a fragmented landscape. Moreover, placing the burden of dataset annotation solely on individual scientists is not a realistic approach.

In this context, I believe it is essential for the working group to:
- Identify the key stakeholders currently active in this space and the data formats they support.
- Assess the advantages and limitations of the principal formats in use today.
- Catalogue the tools and systems that support these formats.

Based on this analysis, the group should then develop informed recommendations regarding formats and tools for future adoption.

I would be glad to contribute to this effort, and I am willing to take a leading role in organizing and steering the working group.

**Comment by SteffenBrinckmann** [comment]
We should acknowledge that the ELNs have different goals with different advantages and disadvantages (like rock-paper-scissor none is superior to all). If the ELNs have different functionalities, they have necessarily different requirements when it comes to exchange format, extractors. ... As such **a exchange-module for all ELN does likely not exist**. We should accept these differences and not force alignment, where it is difficult / impossible.

We also should acknowledge that there are different budgets; some elns have 1 person doing development, design, teaching, ... and on the other hand we have multi-billion euro companies. What can be achieved with one might not be feasible for the other, including participating in all meetings and working groups. Plus there is a difference in data-structure: some that go along the RDF and some that are not; and all 256 shades in between.

I think we aware of xkcd.927 and want to not explicitly create new formats. But that also means that there are already 14 formats and **people chose different versions for different reasons, and that is ok.**

I am aware of 3 ELN exchange formats, in alphabetical order: allotrope (multi-billion company, quite strong RDF), eln (no funding, open consortium, semantic annotation is possible but not necessary) and the openBIS suggestion from Andreas (quite strong RDF). Please correct me if there are more or you do not agree with my categorization.

What triggers me often in these discussions are general statements that focus on the own solution. e.g. "FAIR research data is important and hence we have to use my product XYZ" (I haven't heard this specific argument and do want to not put anybody on the block) because **it implies that other products do not have the same goal.** I think we all have  almost 100% goals but try different paths to get there. We should openly learn, ask: why did you not chose this/that solution. But let's accept  the differences and refrain from such statements.

The open discussion should also lead to an intermediate review: as a community we have been in this RDM-process for ~5 years. Somethings were successful others not (my TAPIR project on extractors). We should try to identify the reasons for success / failure (eg not polished product due to insufficient funding) and try to learn from them. That could lead to some **best practices for RDM project developers**, which I would find really interesting.

Peace.

**Reply by samikanza** [reply]
@SteffenBrinckmann completely agree with your categorisation, and all your points! 

Is anyone interested in doing some hacking using these different approaches? And trying to dig deeper into the actual practicalities of exchanging information between some different ELNs?

**Comment by simontaurus** [comment]
I think its most urgent that we find a common language to express our data models (semantic or not) so we can then focus on the actual mapping of those models.
We are now working for quite some time on a meta-standard to do so, well-placed between Linked Data, Object Oriented Programming and LLM driven approaches: https://github.com/OO-LD/schema

**Reply by edan-bainglass** [reply]
Already stated in several other places here, but @simontaurus good to get your opinion on [RO-Crate Schema Plus](https://github.com/researchobjectschema/ro-crate-interoperability-profile/blob/main/0.2.x/spec.md), and how OO-LD compares with it. Pinging @juan-fuentes-sis and @andreasmeier12

**Reply by SteffenBrinckmann** [reply]
Since the question comes coming up again and again, we created a FAQ about the ability to include semantic description and validation in ro-crate and also .eln.

https://github.com/TheELNConsortium/TheELNFileFormat/blob/master/FAQ.md

Hence the technical solution exists since ro-create exists... See also 
https://github.com/ResearchObject/ro-crate/issues/399

The content issue is that elns have different functions, target-groups and hence different models. Some even allow every user to change the data model. How could you standardize that? Only by agreeing on the most common /basic set. And that is what .eln is doing. 

Everybody is very welcome to join. I look forward to your participation.

**Reply by simontaurus** [reply]
@edan-bainglass : I replied here: https://github.com/MADICES/MADICES-2025/discussions/27#discussioncomment-14801303
